static/js/components/email/hoc.js

```diff
@@ -0,0 +1,116 @@
+        super(props);
+      }
+
+      getActiveEmailType(): ?string {
+        const { email } = this.props;
+        return !isNilOrBlank(email.currentlyActive) ? email.currentlyActive : undefined;
+      }
+
+      getActiveEmailState(): EmailState {
+        const { email } = this.props;
+        return email[email.currentlyActive];
+      }
+
+      openEmailComposer = (emailOpenParams: Object): void => {
+        const { dispatch } = this.props;
+        dispatch(startEmailEdit(emailOpenParams));
+        dispatch(showDialog(EMAIL_COMPOSITION_DIALOG));
+      };
+
+      updateEmailFieldEdit = R.curry((fieldName, e): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        let activeEmail = this.getActiveEmailState();
+        let inputsClone = R.clone(activeEmail.inputs);
+        inputsClone[fieldName] = e.target.value;
+        dispatch(updateEmailEdit({type: currentlyActive, inputs: inputsClone}));
+        if (!R.isEmpty(activeEmail.validationErrors)) {
+          let cloneErrors = emailValidation(inputsClone);
+          dispatch(updateEmailValidation({type: currentlyActive, errors: cloneErrors}));
+        }
+      });
+
+      closeAndClearEmailComposer = (): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        dispatch(clearEmailEdit(currentlyActive));
+        dispatch(hideDialog(EMAIL_COMPOSITION_DIALOG));
+      };
+
+      closeEmailComposerAndSend = (): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        let activeEmail = this.getActiveEmailState();
+        let errors = emailValidation(activeEmail.inputs);
+        dispatch(updateEmailValidation({type: currentlyActive, errors: errors}));
+        if (R.isEmpty(errors)) {
+          dispatch(
+            emailConfigs[currentlyActive].emailSendAction(activeEmail)
+          ).then(() => {
+            this.closeAndClearEmailComposer();
+          });
+        }
+      };
+
+      renderCompositionDialog(): ?React$Element<*> {
+        let { ui: { dialogVisibility } } = this.props;
+
+        let activeEmailType = this.getActiveEmailType();
+
+        if (!activeEmailType) return null;
+
+        return (
+          <EmailCompositionDialog
+            updateEmailFieldEdit={this.updateEmailFieldEdit}
+            closeAndClearEmailComposer={this.closeAndClearEmailComposer}
+            closeEmailComposerAndSend={this.closeEmailComposerAndSend}
+            dialogVisibility={dialogVisibility[EMAIL_COMPOSITION_DIALOG]}
+            activeEmail={this.getActiveEmailState()}
+            title={emailConfigs[activeEmailType].title}
+            subheadingRenderer={emailConfigs[activeEmailType].renderSubheading}
+          />
+        );
+      }
+
+      render() {
+        let emailOpenActions = {};
+        for (let emailType of Object.keys(emailConfigs)) {
+          emailOpenActions[emailType] =
+            emailConfigs[emailType].emailOpenAction.bind(this, emailType);
```

Comment:
proposal: I think it would be a bit cleaner if we could avoid the `this` binding here. We could do so by declaring the `emailOpenAction` functions as curried functions with the following type signature:

```
emailOpenAction :: openEmailComposer -> emailType -> ...args
```

then here we could do something like:

```js
let emailOpenActions = Object.assign({}, ...Object.keys(emailConfigs).map(emailType => ({
  [emailType]: emailConfigs[emailType].emailOpenAction(this.openEmailComposer, emailType)
})));
```

what do you think?